zookeeper-api/src/main/java/org/apache/helix/zookeeper/datamodel/serializer/ZNRecordSerializer.java

```diff
@@ -81,24 +82,44 @@ private static int getListFieldBound(ZNRecord record) {
     serializationConfig.set(SerializationConfig.Feature.AUTO_DETECT_FIELDS, true);
     serializationConfig.set(SerializationConfig.Feature.CAN_OVERRIDE_ACCESS_MODIFIERS, true);
     ByteArrayOutputStream baos = new ByteArrayOutputStream();
-    byte[] serializedBytes;
+    byte[] serializedBytes = new byte[0];
+    boolean isCompressed = false;
+
     try {
       mapper.writeValue(baos, data);
       serializedBytes = baos.toByteArray();
       // apply compression if needed
-      if (record.getBooleanField("enableCompression", false) || serializedBytes.length > ZNRecord.SIZE_LIMIT) {
+      if (ZNRecordUtil.shouldCompress(record, serializedBytes.length)) {
         serializedBytes = GZipCompressionUtil.compress(serializedBytes);
+        isCompressed = true;
       }
     } catch (Exception e) {
-      logger.error("Exception during data serialization. Will not write to zk. Data (first 1k): "
-          + new String(baos.toByteArray()).substring(0, 1024), e);
+      if (serializedBytes.length == 0 || GZipCompressionUtil.isCompressed(serializedBytes)) {
+        serializedBytes = baos.toByteArray();
+      }
+      int firstBytesLength = Math.min(serializedBytes.length, 1024);
+      // TODO: remove logging first N bytes of data to reduce log size.
+      LOG.error("Exception during data serialization. Will not write to zk."
+              + " The first {} bytes of data: {}", firstBytesLength,
+          new String(serializedBytes, 0, firstBytesLength), e);
       throw new ZkClientException(e);
     }
-    if (serializedBytes.length > ZNRecord.SIZE_LIMIT) {
-      logger.error("Data size larger than 1M, ZNRecord.id: " + record.getId()
-          + ". Will not write to zk. Data (first 1k): "
-          + new String(serializedBytes).substring(0, 1024));
-      throw new ZkClientException("Data size larger than 1M, ZNRecord.id: " + record.getId());
+
+    int threshold = ZNRecordUtil.getSerializerThreshold();
```

Comment:
nit: do we want to use long here?

Reply:
I don't think it is necessary because we have a 1 MB as base. I don't think people would set it larger than a max integer. And we do have a check for it. If it is larger than max integer, the property would be parsed to a null and we will use 1 MB.